### ISSUE TITLE
IO::File does not respect open pragma

### ISSUE THREAD

**Issue by Eli-the-Bearded** [Perl/perl5#17727]
This is a bug report for perl from perlbug@eli.users.panix.com,
generated with the help of perlbug 1.40 running under perl 5.24.3.

-----------------------------------------------------------------
[Please describe your issue here]

`perldoc open` says:
    Full-fledged support for I/O layers is now implemented provided Perl
    is configured to use PerlIO as its IO system (which is now the
    default).

    The "open" pragma serves as one of the interfaces to declare default
    "layers" (also known as "disciplines") for all I/O. Any two-argument
    open(), readpipe() (aka qx//) and similar operators found within the
    lexical scope of this pragma will use the declared defaults. Even
    three-argument opens may be affected by this pragma when they don't
    specify IO layers in MODE.

But IO::File / FileHandle `open`s, which are just supposed to be OO
ways to use `open` don't respect the `open` pragma.

```
#!/path/for/-x/mode/perl

use FileHandle;
use open IO => ':utf8';
use IO::File;

my $testfile = '/tmp/utf';
my $testtext = "utf-8\x{2713}\n";

print "My Perl is $]\n";

open(CREATE_TEST, '>', $testfile) || die;
print CREATE_TEST $testtext;
close CREATE_TEST;
printf "Created %s with content <%vx>\n", $testfile, $testtext;

sub testvalue {
  my $type = shift;
  my $value = shift;

  printf "Testing $type open(): <%vx> ", $value;

  if ($value eq $testtext) {
    print "equal\n";
  } else {
    print "NOT equal\n";
  }
}

open(READ_ONE, '<', $testfile) || die;
$_ = <READ_ONE>;
close READ_ONE;
testvalue("old school", $_);

if(open(my $read_two, '<', $testfile)) {
    $_ = <$read_two>;
    testvalue("plain filehandle", $_);
}

my $iofile = IO::File->new();
if ($iofile->open($testfile)) {
    $_ = <$iofile>;
    testvalue("IO::File", $_);
}

my $filehand = new FileHandle;
if ($filehand->open($testfile)) {
    $_ = <$filehand>;
    testvalue("FileHandle", $_);
}
__END__
```

:r! perl-latest -x %
My Perl is 5.024003
Created /tmp/utf with content <75.74.66.2d.38.2713.a>
Testing old school open(): <75.74.66.2d.38.2713.a> equal
Testing plain filehandle open(): <75.74.66.2d.38.2713.a> equal
Testing IO::File open(): <75.74.66.2d.38.e2.9c.93.a> NOT equal
Testing FileHandle open(): <75.74.66.2d.38.e2.9c.93.a> NOT equal

[Please do not change anything below this line]
-----------------------------------------------------------------
---
Flags:
    category=library
    severity=low
---
Site configuration information for perl 5.24.3:

Configured by root at Tue Feb 20 20:46:59 UTC 2018.

Summary of my perl5 (revision 5 version 24 subversion 3) configuration:

  Platform:
    osname=netbsd, osvers=7.1.1, archname=amd64-netbsd-thread-multi
    uname='netbsd trinity.nyc.access.net 7.1.1 netbsd 7.1.1 (panix-xen-std) #1: tue feb 13 19:40:46 est 2018 root@juggler.panix.com:miscobj64miscdevelnetbsd7.1.1srcsysarchamd64compilepanix-xen-std amd64 '
    config_args='-de -A append:ccflags=-DNO_LOCALE_COLLATE -Dusethreads -Duseshrplib'
    hint=previous, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-DNO_LOCALE_COLLATE -fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_FORTIFY_SOURCE=2-DNO_LOCALE_COLLATE',
    optimize='-O',
    cppflags='-DNO_LOCALE_COLLATE -fwrapv -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.8.5', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='ld', ldflags =' -Wl,-rpath,/usr/local/lib -Wl,-rpath,/usr/local/lib -fstack-protector -L/usr/local/lib'
    libpth=/usr/include/gcc-4.8 /usr/lib /usr/local/lib /lib /usr/include/gcc-4.8 /usr/lib
    libs=-lpthread -ldb -lm -lcrypt -lutil -lc -lposix
    perllibs=-lpthread -lm -lcrypt -lutil -lc -lposix
    libc=/lib/libc.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E  -Wl,-R/usr/local/lib/perl5/5.24.3/amd64-netbsd-thread-multi/CORE'
    cccdlflags='-DPIC -fPIC ', lddlflags='-shared  -L/usr/local/lib -fstack-protector'

---
@INC for perl 5.24.3:
    /net/u/3/e/eli/lib
    /net/u/3/e/eli/lib/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.24.3/amd64-netbsd-thread-multi
    /usr/local/lib/perl5/site_perl/5.24.3
    /usr/local/lib/perl5/5.24.3/amd64-netbsd-thread-multi
    /usr/local/lib/perl5/5.24.3

---
Environment for perl 5.24.3:
    HOME=/net/u/3/e/eli
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/net/u/3/e/eli/scripts:/net/u/3/e/eli/bin:/net/u/3/e/eli/contrib:/usr/local/bin:/bin:/usr/bin:/net/u/3/e/eli/usr/bin
    PERL5LIB=/net/u/3/e/eli/lib:/net/u/3/e/eli/lib/site_perl/5.6.1
    PERL_BADLANG=0
    SHELL=/usr/local/bin/ksh93

**Comment by Grinnz** [comment]
This is a duplicate of https://github.com/Perl/perl5/issues/12749

**Comment by jkeenan** [comment]
> This is a duplicate of #12749

Yes, this is a duplicate of 12749.  However, discussion on that ticket petered out inconclusively in 2013.  We probably should have simply made documentation improvements somewhere.

Would doc improvements suffice in the current case?  If so, where should we make them?

Thank you very much.
Jim Keenan

**Comment by Eli-the-Bearded** [comment]
Well the documentation I quoted from `perldoc open`, for starters, is rather misleading vis-a-vis IO::File. It would be one thing if IO::File were just some random module, but it's not, it's part of the core modules.

**Comment by tonycoz** [comment]
This is related to #17458 too.  That was resolved by updating the documentation for `-C`.
